groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -218,16 +220,319 @@ extern const Optional_OptNoSuchType optNoSuchType;
+};
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+template <class TYPE, class OPT_TYPE>
+struct Optional_AssignsFrom
+: bsl::integral_constant<
+      bool,
+      std::is_assignable<TYPE&, const OPT_TYPE&>::value ||
+          std::is_assignable<TYPE&, OPT_TYPE&>::value ||
+          std::is_assignable<TYPE&, const OPT_TYPE>::value ||
+          std::is_assignable<TYPE&, OPT_TYPE>::value> {
+};
 #else
-#define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE bsl::is_trivially_copyable
-    // C++03 does not provide a trivially destructible trait.  Instead we use
-    // 'bsl::is_trivially_copyable' which implies the type is also trivially
-    // destructible.
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFrom : bsl::integral_constant<bool, false> {
+    // We only use '|| BloombergLP::bslstl::Optional_AssignsFrom' in
+    // 'bsl::optional' constraints. In order to ignore
+    // Optional_AssignsFromOptional trait in C++03, we set it to false so it
+    // never affects the trait it appears in.
+};
 #endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
 
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromBslOptional
+: bsl::integral_constant<
+      bool,
+      Optional_ConvertsFrom<TYPE, bsl::optional<ANY_TYPE> >::value> {
+};
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromBslOptional
+: bsl::integral_constant<
+      bool,
+      Optional_AssignsFrom<TYPE, bsl::optional<ANY_TYPE> >::value> {
+};
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromStdOptional
+: bsl::integral_constant<
+      bool,
+      Optional_ConvertsFrom<TYPE, std::optional<ANY_TYPE> >::value> {
+};
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromStdOptional
+: bsl::integral_constant<
+      bool,
+      Optional_AssignsFrom<TYPE, std::optional<ANY_TYPE> >::value> {
+};
+
+#else
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromStdOptional : bsl::integral_constant<bool, false> {
+};
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromStdOptional : bsl::integral_constant<bool, false> {
+};
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromOptional
+: bsl::integral_constant<
+      bool,
+      Optional_ConvertsFromBslOptional<TYPE, ANY_TYPE>::value ||
+          Optional_ConvertsFromStdOptional<TYPE, ANY_TYPE>::value> {
+};
+
+template <class TYPE, class ANY_TYPE>
+struct Optional_AssignsFromOptional
+: bsl::integral_constant<
+      bool,
+      Optional_AssignsFromBslOptional<TYPE, ANY_TYPE>::value ||
+          Optional_AssignsFromStdOptional<TYPE, ANY_TYPE>::value> {
+};
+
+// Remove CV qualifiers and references from type 'U'
+template <class TYPE>
+struct Optional_RemoveCVRef {
+    typedef typename bsl::remove_cv<bsl::remove_reference<TYPE> >::type type;
```

Comment:
```suggestion
    typedef typename bsl::remove_cv<bsl::remove_reference<TYPE>::type>::type type;
```

Reply:
already fixed as a part of a different change, leaving the comment open until I implement the test for Optional_RemoveCVRef